v2/internal/parse.go

```diff
@@ -39,49 +40,103 @@ func ParseMessageFileBytes(buf []byte, path string, unmarshalFuncs map[string]Un
 		}
 	}
+	var err error
 	var raw interface{}
-	if err := unmarshalFunc(buf, &raw); err != nil {
+	if err = unmarshalFunc(buf, &raw); err != nil {
 		return nil, err
 	}
+
+	if messageFile.Messages, err = recGetMessages(raw, isMessage(raw), true); err != nil {
+		return nil, err
+	}
+
+	return messageFile, nil
+}
+
+const nestedSeparator = "."
+
+// recGetMessages looks for translation messages inside "raw" parameter,
+// scanning nested maps using recursion.
+func recGetMessages(raw interface{}, isMapMessage, isInitialCall bool) ([]*Message, error) {
+	var messages []*Message
+	var err error
+
 	switch data := raw.(type) {
+	case string:
+		if isInitialCall {
+			return nil, errors.New("invalid translation file, expected key-values, got a single value")
+		}
+		m, err := NewMessage(data)
+		return []*Message{m}, err
+
 	case map[string]interface{}:
-		messageFile.Messages = make([]*Message, 0, len(data))
-		for id, data := range data {
+		if isMapMessage {
 			m, err := NewMessage(data)
+			return []*Message{m}, err
+		}
+		messages = make([]*Message, 0, len(data))
+		for id, data := range data {
+			// recursively scan map items
+			messages, err = addChildMessages(id, data, messages)
 			if err != nil {
 				return nil, err
 			}
-			m.ID = id
-			messageFile.Messages = append(messageFile.Messages, m)
 		}
+
 	case map[interface{}]interface{}:
-		messageFile.Messages = make([]*Message, 0, len(data))
+		if isMapMessage {
+			m, err := NewMessage(data)
+			return []*Message{m}, err
+		}
+		messages = make([]*Message, 0, len(data))
 		for id, data := range data {
 			strid, ok := id.(string)
 			if !ok {
 				return nil, fmt.Errorf("expected key to be string but got %#v", id)
 			}
-			m, err := NewMessage(data)
+			// recursively scan map items
+			messages, err = addChildMessages(strid, data, messages)
 			if err != nil {
 				return nil, err
 			}
-			m.ID = strid
-			messageFile.Messages = append(messageFile.Messages, m)
 		}
+
 	case []interface{}:
 		// Backward compatibility for v1 file format.
-		messageFile.Messages = make([]*Message, 0, len(data))
+		messages = make([]*Message, 0, len(data))
 		for _, data := range data {
-			m, err := NewMessage(data)
+			// recursively scan slice items
+			childMessages, err := recGetMessages(data, isMessage(data), false)
 			if err != nil {
 				return nil, err
 			}
-			messageFile.Messages = append(messageFile.Messages, m)
+			messages = append(messages, childMessages...)
 		}
+
 	default:
 		return nil, fmt.Errorf("unsupported file format %T", raw)
 	}
-	return messageFile, nil
+
+	return messages, nil
+}
+
+func addChildMessages(id string, data interface{}, messages []*Message) ([]*Message, error) {
+	isChildMessage := isMessage(data)
+	childMessages, err := recGetMessages(data, isChildMessage, false)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range childMessages {
+		if isChildMessage {
+			if m.ID == "" {
+				m.ID = id // start with innermost key
+			}
```

Comment:
Can you explain to me when the else case happens here? I am just trying to follow the logic.

Reply:
Indeed, I did not get this line correct from the start, it is not so obvious :-)

All messages created by `recGetMessages()` use `NewMessage()` to setup message content and fieds. When `NewMessage()` returns, it sets the message fields it finds inside the message submap: if `id` was defined, then it is set, otherwise it is not and we need to set it manually with the map key.

In practice, this is tested with 2 tests:
* "basic test" (`{"hello": "world"}`) creates a message with `Other` content `"world"`, but no `ID`, so we manually set it to `"hello"` when iterating through the high-level map (line 132).
* "basic test with id" (`{"key": {"id": "forced.id"}}`) creates a message with no content, but with `ID: "forced.id"`. In this case, we do not want to reset the `ID` to `"key"` (passes test line 130, but not the one line 131).